### ISSUE TITLE
Progressive discount based on number of items purchased PDP

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#764]
_Imported from trac ticket 764.
Created by **kbower**
Opened in trac: Mar 2, 2011, 11:28:34 AM
Last modified in trac: Mar 18, 2011, 2:39:09 PM_
***
Progressive discount based on number of items purchased   PDP needed by 03/07/11

a.       10% off any one item when you buy one item or

b.      15% off any one item when you buy two items 0r

c.       20% off any one item when you buy three or more items

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 3, 2011, 9:00:53 AM:_

Changeset [[1254]](../commit/ff82bd2dcbe5674b53d9cd63a441e4d2f7d09cd3)

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 3, 2011, 9:37:41 AM:_

I've got the "Progressive discount based on number of items purchased PDP needed by 03/07/11" mostly implemented, but I was wondering if the same exclusions for "number of items purchased" applies as applied with the Presidents Sale (the number of items excludes certain minor codes)?

Also, when rereading this email (below) I realize I failed to recognize the second part of #2 below:  "should only include inventory items and exclude warranties and fabric protection."
In other words, if they reached $5000 only because they purchased warranties/fab, the system counted that and gave them 50% off an item.

My apologies for that.  I guess we should leave it implemented the same way it is for historical editing of orders, but is that a requirement for the upcoming promotions?  Many of them refer to "total purchase of".   Does that really mean total purchase or does it mean total purchase excluding certain items, such as FAB, etc.?

Kent

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 3, 2011, 12:04:03 PM:_

Kent,

Yes, the same exclusions for “Progressive discount” applies when determining the count of items sold.

Thanks,

*Wayne Liburd*

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 3, 2011, 12:39:03 PM:_

Ok, let me make sure I've got this straight:

*There are two separate exclusion lists:*

When determining which lines count towards "number of items purchased":

ITM.ITM_TP_CD = 'INV' and

ITM.INVENTORY = 'Y' and

SO_LN.UNIT_PRC > 0 and

        ITM.MNR_CD NOT IN ('097', '390', '391', '392', '393', '395', '513', '523', '561', '565', '570', '700', '710', '720', '730', '740', '790', '900')

When determining which lines add to the "purchase total":
ITM.INVENTORY = 'Y' and ITM.MNR_CD NOT IN ('097', '390', '391', '392', '393', '395', '513', '523', '561', '565', '570', '900').


And this is for all the discounts going forward that reference "purchase total" or "number of items purchased", correct?

**Comment by kentbower** [comment]
_**wayne** trac comment on Mar 3, 2011, 3:01:44 PM:_

Kent,

 

Correct.

 

Thanks,

 

Wayne Liburd

Director of MIS

Kanes Furniture Corporation

5700 70th Ave North

Pinellas Park FL 33781

Tel: +1 (727) 545-9555 ext. 350

Fax: +1 (727) 541-6960

From: Kent Bower [mailto:kb@retailarchitects.com] 
Sent: Thursday, March 03, 2011 12:38 PM
To: Wayne Liburd
Cc: Rob Neville; issues@retailarchitects.com
Subject: #764:

 

Ok, let me make sure I've got this straight:

There are two separate exclusion lists:

When determining which lines count towards "number of items purchased":

ITM.ITM_TP_CD = 'INV' and

ITM.INVENTORY = 'Y' and

SO_LN.UNIT_PRC > 0 and

        ITM.MNR_CD NOT IN ('097', '390', '391', '392', '393', '395',
'513', '523', '561', '565', '570', '700', '710', '720', '730', '740',
'790', '900')

When determining which lines add to the "purchase total":
    ITM.INVENTORY = 'Y' and ITM.MNR_CD NOT IN ('097', '390', '391',
'392', '393', '395', '513', '523', '561', '565', '570', '900').


And this is for all the discounts going forward that reference "purchase
total" or "number of items purchased", correct?

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 7, 2011, 9:58:51 AM:_

(Changeset [[1271]](../commit/e3c744ff1f749c9313cdbc990ab194f69fb040cf))
 * last minute changes as decribed by Kane's for PDP discount, see #764

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 7, 2011, 10:04:14 AM:_

Need refactor [r1271](../commit/e3c744ff1f749c9313cdbc990ab194f69fb040cf)

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 7, 2011, 12:22:47 PM:_

(Changeset [[1275]](../commit/2d67cefe01b5a926ba6e158f21619d09433b1494))
 * Kane's PDP discount (closes #764)

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 9, 2011, 4:46:40 PM:_

The following items need to be excluded from any discounting by all the discount codes requested in addition to what was given to you previously:

1.       Exclude any skus where ITM.DROP_CD = ‘S’
2.       Exclude any skus where ITM.VE_CD IN (‘TPED’, ‘STEA’)
3.       Exclude where ITM.VE_CD = ‘KING’ AND ITM.FAMILY_CD = ‘58455’
4.       Exclude where ITM.VE_CD = ‘SERT’ AND ITM.FAMILY_CD IN (‘32042’, ‘31783’) AND ITM.CAT_CD BETWEEN ‘506’ and ‘511’
5.       Exclude where ITM_CD IN (SELECT ITM_CD FROM ITM$PROM_PRC_TP WHERE PRC_CD = ‘RT’)
6.       Exclude asis (WHERE SO.SHIP_TO_ZONE_CD = ‘FPU’ AND SO.PU_DEL = ‘P’ AND SO.PU_DEL_STORE_CD <> ‘00’)
7.       Exclude where ITM.MNR_CD IN (‘700’, ‘710’, ‘720’, ‘730’, ‘740’, ‘790’)

**[EDIT] Updated specs here https://trac.retailarchitects.com:444/trac/ticket/765#comment:11**

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 9, 2011, 4:47:37 PM:_

Afternoon Wayne,
A few questions and concerns:
 * By when? You asked for PDP promotion by Monday and I don't know the actual date you need it by.
 * Will this apply to PDP?
 * We've talked about the dangers of redefining an existing discount.  Not only for historical reasons, but if an order with such a discount is opened and resaved the discount will then be calculated differently and the total sale amount changed when the order is saved.  Do these rules apply to all discounts, including GERS discounts and/or already existing discounts (PRS,SEN, etc) or only discounts you haven't run yet?
 * I specifically asked Daniel about whether a discount should be allowed in addition to any promotions and he specifically said "yes".  In fact, I think he may have even sited the Red Tag promotion.  So I just want to verify that concerning #5.
----
Kent,

None of the discount I sent you have started. The Senior Sale starts tomorrow and I guess because the definition is different we will have to use a different code. Let’s use ‘SNR’. The President Day Sale ‘PRS’ will also need a new code for next year’s sale. This will also apply to the PDP code with starts on Thursday. This was caught during the testing today. The advertised disclaimers which I did not have at the time I sent you the definition of the promotion does include everything I sent you including excluding Red Tag items. One of the reasons I have you set it up earlier in addition to getting some testing done is because we usually honor the promotion a day or two early if customers come in because they have received the promotion in the mail however, this has not been the case so far.

Thanks,

Wayne Liburd

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 9, 2011, 4:53:08 PM:_

**6. Was originally _Exclude where SO_LN.LOC_CD = STORE_CD_**
----
# 6 is impossible to implement, unless you change it to checking for the presence of SKU 999999999 or fromsiteid != '00' (not sure if that would work, something with P instead of D)...
(Lines are not filled until well after the prices are calculated, etc.)

Kent
----
Kent,

Yes, the seven exclusions apply to the senior discount.

How about if SO_LN.UNIT_PRC is less than the lessor of the retail price or the promotion price or package split out price for AS IS items?

Thanks,

Wayne
----
That won't account for package substitution which has changed the price. 
----
Kent,

You can use the same logic for packages and exculde all components skus if the package price or sum of the components is less.

Wayne
----
[Same problem.]
----
If the PU_DEL is 'P' and PU_DEL_STORE_CD is not '00', is that enough to know it doesn't qualify for the discount?

Kent
----
Kent,
 Yes, that is enough.

Thanks,
Wayne Liburd
----
If that is always correct, we could use those criteria for the receipt as well, or is the disclaimer printing at the right times?
----
Kent,
What logic are you using now?
----
For the receipt it is using the fill location and fill siteid.

It is also checking the shipzoneid of 'FPU'.... is that all by itself enough? 
Do people ever get floor model's delivered (I imagine yes)?

Here is the function's code for the receipt disclaimer:
def kanes_asispiece(orderdetail):
    if not orderdetail.filllocationid:
        return False
    if orderdetail.fillsiteid != u'00' and orderdetail.filllocationid == u'ACC':
        return True
    elif orderdetail.order.shipzoneid==u'FPU':
        return True
    return False

I'm not sure about the PU_DEL is 'P' and PU_DEL_STORE_CD is not '00' logic.....  can you give that another look?

Kent
----
Kent,
Generally, someone purchasing the floor model will pick it up or the store will arrange for local delivery.

For both printing the receipt disclaimer and excluding from the discount we should use the following:

WHERE SO.SHIP_TO_ZONE_CD = ‘FPU’
AND       SO.PU_DEL = ‘P’
AND       SO.PU_DEL_STORE_CD <> ‘00’

Thanks,
Wayne Liburd

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 9, 2011, 4:59:28 PM:_

## Red Tag exclusion
Kent,

Another note, the Red tag program has a beginning date of 01/01/50 and an ending date of 12/31/49 so it is always on and skus are just added to the promotion do the date logic does not help us. We seem to need a log of the skus, store and price that were active at that time. However, the date logic will become important for other promotions but it need to combined with a table that logs that historical information.
----
Once a SKU goes on the Red tag list, does it ever get removed?

Kent
----
Kent,

It rarely gets removed but it is possible.
----
If starting now and going forward, you never remove SKUs from the RT promotion, we can use your original suggestion which is this:

5.       Exclude where ITM_CD IN (SELECT ITM_CD FROM ITM$PROM_PRC_TP WHERE PRC_CD = ‘RT’)

Do you agree and can we make that the rule: "Once Red Tag always Red Tag" ?

Kent
----
Kent,

I agree we can make that rule.

Thanks,
Wayne Liburd

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 9, 2011, 5:00:59 PM:_

(Changeset [[1290]](../commit/e3905ed295f16df7cf07be932273675d1161ef25))
 * change receipt "as is" disclaimer logic: instead of filllocation, print when "order.shipzoneid == u'FPU' and order.deliverytype == u'P' and order.fromsiteid != u'00'"
 * implement 'SNR' and 'PDP' discounts including 7 product exclusion criteria. (closes #764)

**Comment by kentbower** [comment]
_**kbower** trac status change on Mar 9, 2011, 5:00:59 PM:_ **closed**

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 10, 2011, 1:58:24 PM:_

Replying to [kbower](comment:16):
> '''6. Was originally ''Exclude where SO_LN.LOC_CD = STORE_CD'''''
> 
> For both printing the receipt disclaimer and excluding from the discount we should use the following:
> 
> WHERE SO.SHIP_TO_ZONE_CD = ‘FPU’
> AND       SO.PU_DEL = ‘P’
> AND       SO.PU_DEL_STORE_CD <> ‘00’
> 
> Thanks,
> Wayne Liburd


Kent,

We need to add one more criteria where the (SO_LN.UNIT_PRC < ITM.RET_PRC) or if there is a promotion (SO_LN.UNIT_PRC < active promotion calendar price)

Thanks,

Wayne
----
I spoke with Wayne and we've determined that the presence of orderdetail.priceauthorizedby (SO_LN.PRC_CHG_APP_CD) is sufficient evidence that the price has been lowered.  This is not 100% accurate, but Kane's is fine with it as a rule.

**Comment by kentbower** [comment]
_**kbower** trac status change on Mar 10, 2011, 1:58:59 PM:_ **reopened**

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 10, 2011, 3:19:59 PM:_

Replying to [kbower](comment:19):
> Replying to [comment:16 kbower]:
> > '''6. Was originally ''Exclude where SO_LN.LOC_CD = STORE_CD'''''
> > 
> > For both printing the receipt disclaimer and excluding from the discount we should use the following:
> > 
> > WHERE SO.SHIP_TO_ZONE_CD = ‘FPU’
> > AND       SO.PU_DEL = ‘P’
> > AND       SO.PU_DEL_STORE_CD <> ‘00’
> > 
> > Thanks,
> > Wayne Liburd
> 
> I spoke with Wayne and we've determined that the presence of orderdetail.priceauthorizedby (SO_LN.PRC_CHG_APP_CD) is sufficient evidence that the price has been lowered.  This is not 100% accurate, but Kane's is fine with it as a rule.


[Kent:]

 * In addition to checking SO_LN.PRC_CHG_APP_CD, should we also check that the product is INVENTORY='Y'?
 * Anything else that would be needed ?
----
Kent,
Yes, we should. I thought we had that already. I think that should cover it.
Thanks,
Wayne

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 10, 2011, 3:23:36 PM:_

(Changeset [[1306]](../commit/fe55360924282ff91635f0d17b23f58dcfdccd1c))
 * Only identify an item as "As is" if a price change has (ever) been authorized for it.  (Fixes #764, again)

**Comment by kentbower** [comment]
_**kbower** trac status change on Mar 10, 2011, 3:23:36 PM:_ **closed**
